File: tests/format.test.ts
import { describe, test, expect } from "bun:test";
import { formatTable, formatJson, type WeeklySummary } from "../src/format.ts";

const sampleSummary: WeeklySummary = {
  weekStart: "2026-02-09",
  weekEnd: "2026-02-13",
  clients: [
    { name: "Acme Corp", daily: [2.0, 3.5, 4.0, 2.0, 1.5], total: 13.0 },
    { name: "Globex Inc", daily: [1.0, 0, 2.0, 3.0, 2.0], total: 8.0 },
  ],
  grandTotal: 21.0,
};

describe("formatTable", () => {
  test("contains week header", () => {
    const output = formatTable(sampleSummary);
    expect(output).toContain("Week of Feb 9 – Feb 13, 2026");
  });

  test("contains column headers", () => {
    const output = formatTable(sampleSummary);
    expect(output).toContain("Client");
    expect(output).toContain("Mon");
    expect(output).toContain("Fri");
    expect(output).toContain("Total");
  });

  test("contains client rows with hours", () => {
    const output = formatTable(sampleSummary);
    expect(output).toContain("Acme Corp");
    expect(output).toContain("13.0h");
    expect(output).toContain("Globex Inc");
    expect(output).toContain("8.0h");
  });

  test("shows dash for zero hours", () => {
    const output = formatTable(sampleSummary);
    // Globex has 0 on Tuesday
    expect(output).toContain("—");
  });

  test("contains grand total", () => {
    const output = formatTable(sampleSummary);
    const lines = output.split("\n");
    const totalLine = lines.find((l) => l.startsWith("Total"));
    expect(totalLine).toBeDefined();
    expect(totalLine).toContain("21.0h");
  });

  test("renders empty week correctly", () => {
    const empty: WeeklySummary = {
      weekStart: "2026-02-09",
      weekEnd: "2026-02-13",
      clients: [],
      grandTotal: 0,
    };
    const output = formatTable(empty);
    expect(output).toContain("Week of Feb 9 – Feb 13, 2026");
    expect(output).toContain("Total");
    expect(output).toContain("—h");
  });
});

describe("formatJson", () => {
  test("outputs valid JSON", () => {
    const output = formatJson(sampleSummary);
    const parsed = JSON.parse(output);
    expect(parsed).toBeDefined();
  });

  test("preserves structure", () => {
    const output = formatJson(sampleSummary);
    const parsed = JSON.parse(output);
    expect(parsed.weekStart).toBe("2026-02-09");
    expect(parsed.weekEnd).toBe("2026-02-13");
    expect(parsed.clients).toHaveLength(2);
    expect(parsed.grandTotal).toBe(21.0);
  });

  test("includes client details", () => {
    const output = formatJson(sampleSummary);
    const parsed = JSON.parse(output);
    expect(parsed.clients[0].name).toBe("Acme Corp");
    expect(parsed.clients[0].daily).toEqual([2.0, 3.5, 4.0, 2.0, 1.5]);
    expect(parsed.clients[0].total).toBe(13.0);
  });

  test("handles empty week", () => {
    const empty: WeeklySummary = {
      weekStart: "2026-02-09",
      weekEnd: "2026-02-13",
      clients: [],
      grandTotal: 0,
    };
    const output = formatJson(empty);
    const parsed = JSON.parse(output);
    expect(parsed.clients).toEqual([]);
    expect(parsed.grandTotal).toBe(0);
  });
});

File: internal/format/format_test.go
package format

import (
	"encoding/json"
	"strings"
	"testing"
)

func TestFormatHours(t *testing.T) {
	tests := []struct {
		input float64
		want  string
	}{
		{0, "—"},
		{1.0, "1.0"},
		{8.5, "8.5"},
		{0.5, "0.5"},
		{10.25, "10.2"}, // rounds to 1 decimal
	}
	for _, tt := range tests {
		got := formatHours(tt.input)
		if got != tt.want {
			t.Errorf("formatHours(%v) = %q, want %q", tt.input, got, tt.want)
		}
	}
}

func TestFormatDateRange(t *testing.T) {
	got := formatDateRange("2026-02-23", "2026-02-27")
	want := "Feb 23 – Feb 27, 2026"
	if got != want {
		t.Errorf("formatDateRange = %q, want %q", got, want)
	}
}

func TestTable(t *testing.T) {
	summary := &WeeklySummary{
		WeekStart:  "2026-02-23",
		WeekEnd:    "2026-02-27",
		GrandTotal: 24.0,
		Clients: []ClientSummary{
			{Name: "Acme Corp", Daily: []float64{8.0, 8.0, 0, 0, 0}, Total: 16.0},
			{Name: "Beta Inc", Daily: []float64{0, 0, 4.0, 4.0, 0}, Total: 8.0},
		},
	}

	result := Table(summary)

	// Check structural elements
	if !strings.Contains(result, "Week of Feb 23 – Feb 27, 2026") {
		t.Error("missing date range header")
	}
	if !strings.Contains(result, "Client") {
		t.Error("missing Client column header")
	}
	for _, day := range []string{"Mon", "Tue", "Wed", "Thu", "Fri"} {
		if !strings.Contains(result, day) {
			t.Errorf("missing day header: %s", day)
		}
	}
	if !strings.Contains(result, "Total") {
		t.Error("missing Total column")
	}
	if !strings.Contains(result, "Acme Corp") {
		t.Error("missing client name Acme Corp")
	}
	if !strings.Contains(result, "Beta Inc") {
		t.Error("missing client name Beta Inc")
	}
	// Zero hours should show dash
	if !strings.Contains(result, "—") {
		t.Error("zero hours should display as dash")
	}
	// Separator lines
	if !strings.Contains(result, "─") {
		t.Error("missing separator line")
	}
}

func TestTableTruncatesLongNames(t *testing.T) {
	summary := &WeeklySummary{
		WeekStart:  "2026-02-23",
		WeekEnd:    "2026-02-27",
		GrandTotal: 8.0,
		Clients: []ClientSummary{
			{Name: "Very Long Client Name That Exceeds Twenty Chars", Daily: []float64{8.0, 0, 0, 0, 0}, Total: 8.0},
		},
	}

	result := Table(summary)
	// Name should be truncated to 20 chars
	if strings.Contains(result, "Very Long Client Name That Exceeds Twenty Chars") {
		t.Error("long client name should be truncated")
	}
}

func TestJSON(t *testing.T) {
	summary := &WeeklySummary{
		WeekStart:  "2026-02-23",
		WeekEnd:    "2026-02-27",
		GrandTotal: 8.0,
		Clients: []ClientSummary{
			{Name: "Acme Corp", Daily: []float64{8.0, 0, 0, 0, 0}, Total: 8.0},
		},
	}

	result := JSON(summary)

	// Should be valid JSON
	var parsed WeeklySummary
	if err := json.Unmarshal([]byte(result), &parsed); err != nil {
		t.Fatalf("JSON output is not valid JSON: %v", err)
	}

	// Verify round-trip
	if parsed.WeekStart != "2026-02-23" {
		t.Errorf("weekStart = %q, want %q", parsed.WeekStart, "2026-02-23")
	}
	if parsed.GrandTotal != 8.0 {
		t.Errorf("grandTotal = %v, want 8.0", parsed.GrandTotal)
	}
	if len(parsed.Clients) != 1 {
		t.Fatalf("clients len = %d, want 1", len(parsed.Clients))
	}
	if parsed.Clients[0].Name != "Acme Corp" {
		t.Errorf("client name = %q, want %q", parsed.Clients[0].Name, "Acme Corp")
	}
}
